File: src/kubectl-tools.ts
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js"
import { kubectlContextArgs, namespaceArgs } from "./args.js"
import {
  KubectlAuthCanIInput,
  KubectlDescribeInput,
  KubectlEventsInput,
  KubectlGetInput,
  KubectlLogsInput
} from "./schemas.js"
import { commandTool } from "./tool-runner.js"
import type { RuntimeConfig } from "./types.js"

export function registerKubectlTools(server: McpServer, config: RuntimeConfig): void {
  server.registerTool(
    "kubectl_get",
    { description: "Run a read-only kubectl get with an explicit EKS context.", inputSchema: KubectlGetInput.shape },
    ({ cluster, resource, name, namespace, allNamespaces, output, selector }) =>
      commandTool(config, {
        command: "kubectl",
        args: [
          ...kubectlContextArgs(cluster),
          "get",
          resource,
          ...(name === undefined ? [] : [name]),
          ...(allNamespaces ? ["--all-namespaces"] : namespaceArgs(namespace)),
          "--output",
          output,
          ...(selector === undefined ? [] : ["--selector", selector])
        ]
      })
  )

  server.registerTool(
    "kubectl_describe",
    { description: "Run kubectl describe with an explicit EKS context.", inputSchema: KubectlDescribeInput.shape },
    ({ cluster, resource, name, namespace }) =>
      commandTool(config, {
        command: "kubectl",
        args: [...kubectlContextArgs(cluster), "describe", resource, name, ...namespaceArgs(namespace)]
      })
  )

  server.registerTool(
    "kubectl_logs",
    { description: "Read pod logs with bounded tail output.", inputSchema: KubectlLogsInput.shape },
    ({ cluster, pod, namespace, container, tail, since }) =>
      commandTool(config, {
        command: "kubectl",
        args: [
          ...kubectlContextArgs(cluster),
          "logs",
          pod,
          ...namespaceArgs(namespace),
          "--tail",
          String(tail),
          ...(container === undefined ? [] : ["--container", container]),
          ...(since === undefined ? [] : ["--since", since])
        ]
      })
  )

  server.registerTool(
    "kubectl_events",
    { description: "List events sorted by creation time.", inputSchema: KubectlEventsInput.shape },
    ({ cluster, namespace, fieldSelector }) =>
      commandTool(config, {
        command: "kubectl",
        args: [
          ...kubectlContextArgs(cluster),
          "get",
          "events",
          ...namespaceArgs(namespace),
          "--sort-by",
          ".metadata.creationTimestamp",
          ...(fieldSelector === undefined ? [] : ["--field-selector", fieldSelector])
        ]
      })
  )

  server.registerTool(
    "kubectl_auth_can_i",
    { description: "Check Kubernetes RBAC for a verb/resource pair.", inputSchema: KubectlAuthCanIInput.shape },
    ({ cluster, namespace, verb, resource }) =>
      commandTool(config, {
        command: "kubectl",
        args: [...kubectlContextArgs(cluster), "auth", "can-i", verb, resource, ...namespaceArgs(namespace)]
      })
  )
}

File: safe_eks_mcp/kubectl_tools.py
from __future__ import annotations

from typing import Literal

from .args import kubectl_context_args, namespace_args
from .runner import command_tool
from .types import CliInvocation, RuntimeConfig, ToolPayloadJson
from .validation import (
    optional_namespace,
    optional_since,
    require_cluster,
    require_name,
    require_region,
    require_resource,
)

OutputFormat = Literal["json", "yaml", "wide", "name"]


def kubectl_get(
    config: RuntimeConfig,
    region: str,
    cluster: str,
    resource: str,
    name: str | None = None,
    namespace: str | None = None,
    allNamespaces: bool = False,
    output: OutputFormat = "json",
    selector: str | None = None,
) -> ToolPayloadJson:
    _ = require_region(region)
    args = [*kubectl_context_args(require_cluster(cluster)), "get", require_resource(resource)]
    if name is not None:
        args.append(require_name(name, "name"))
    args.extend(("--all-namespaces",) if allNamespaces else namespace_args(optional_namespace(namespace)))
    args.extend(("--output", output))
    if selector is not None:
        args.extend(("--selector", selector))
    return command_tool(config, CliInvocation(command="kubectl", args=tuple(args)))


def kubectl_describe(
    config: RuntimeConfig, region: str, cluster: str, resource: str, name: str, namespace: str | None = None
) -> ToolPayloadJson:
    _ = require_region(region)
    return command_tool(
        config,
        CliInvocation(
            command="kubectl",
            args=(
                *kubectl_context_args(require_cluster(cluster)),
                "describe",
                require_resource(resource),
                require_name(name, "name"),
                *namespace_args(optional_namespace(namespace)),
            ),
        ),
    )


def kubectl_logs(
    config: RuntimeConfig,
    region: str,
    cluster: str,
    pod: str,
    namespace: str | None = None,
    container: str | None = None,
    tail: int = 200,
    since: str | None = None,
) -> ToolPayloadJson:
    _ = require_region(region)
    bounded_tail = min(max(tail, 1), 10_000)
    args = [
        *kubectl_context_args(require_cluster(cluster)),
        "logs",
        require_name(pod, "pod"),
        *namespace_args(optional_namespace(namespace)),
        "--tail",
        str(bounded_tail),
    ]
    if container is not None:
        args.extend(("--container", require_name(container, "container")))
    if since is not None:
        args.extend(("--since", optional_since(since) or since))
    return command_tool(config, CliInvocation(command="kubectl", args=tuple(args)))


def kubectl_events(
    config: RuntimeConfig, region: str, cluster: str, namespace: str | None = None, fieldSelector: str | None = None
) -> ToolPayloadJson:
    _ = require_region(region)
    args = [
        *kubectl_context_args(require_cluster(cluster)),
        "get",
        "events",
        *namespace_args(optional_namespace(namespace)),
        "--sort-by",
        ".metadata.creationTimestamp",
    ]
    if fieldSelector is not None:
        args.extend(("--field-selector", fieldSelector))
    return command_tool(config, CliInvocation(command="kubectl", args=tuple(args)))


def kubectl_auth_can_i(
    config: RuntimeConfig, region: str, cluster: str, verb: str, resource: str, namespace: str | None = None
) -> ToolPayloadJson:
    _ = require_region(region)
    return command_tool(
        config,
        CliInvocation(
            command="kubectl",
            args=(
                *kubectl_context_args(require_cluster(cluster)),
                "auth",
                "can-i",
                require_name(verb, "verb"),
                require_resource(resource),
                *namespace_args(optional_namespace(namespace)),
            ),
        ),
    )
